MastodonSDK/Sources/MastodonUI/Helper/MediaLayoutHelper.swift

```diff
@@ -0,0 +1,314 @@
+				var h: Float = min(maxHeight-hCover-gap, min(w2/ratios[1], w2/ratios[2]))
+				if hCover+h < minHeight {
+					let prevTotalHeight = hCover+h
+					hCover = minHeight*(hCover/prevTotalHeight)
+					h = minHeight*(h/prevTotalHeight)
+				}
+				
+				return MediaLayoutResult(width: Int(maxWidth),
+										 height: Int((hCover+h+gap).rounded()),
+										 columnSizes: [Int(w2.rounded()), Int(maxWidth-w2.rounded())],
+										 rowSizes: [Int(hCover.rounded()), Int(h.rounded())],
+										 tiles: [
+											MediaLayoutResult.Tile(colSpan: 2, rowSpan: 1, startCol: 0, startRow: 0),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 0, startRow: 1),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 1, startRow: 1)
+										 ])
+			} else {
+				// One on the left, two smaller ones on the right
+				let height: Float = min(maxHeight, maxWidth*0.66/avgRatio)
+				let wCover: Float = min(height*ratios[0], (maxWidth-gap)*0.66)
+				let h1: Float = ratios[1]*(height-gap)/(ratios[2]+ratios[1])
+				let h0: Float = height-h1-gap
+				let w: Float = min(maxWidth-wCover-gap, h1*ratios[2], h0*ratios[1])
+				
+				return MediaLayoutResult(width: Int((wCover+w+gap).rounded()),
+										 height: Int(height.rounded()),
+										 columnSizes: [Int(wCover.rounded()), Int(w.rounded())],
+										 rowSizes: [Int(h0.rounded()), Int(h1.rounded())],
+										 tiles: [
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 2, startCol: 0, startRow: 0),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 1, startRow: 0),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 1, startRow: 1)
+										 ])
+			}
+		case 4:
+			if ratios[0]>1.2*maxRatio || avgRatio>1.5*maxRatio || allAreWide {
+				// One above three smaller ones
+				var hCover: Float = min(maxWidth/ratios[0], (maxHeight-gap)*0.66)
+				var h: Float = (maxWidth-2.0*gap)/(ratios[1]+ratios[2]+ratios[3])
+				let w0: Float = h*ratios[1]
+				let w1: Float = h*ratios[2]
+				h = min(maxHeight-hCover-gap, h)
+				if hCover+h<minHeight {
+					let prevTotalHeight = hCover+h
+					hCover = minHeight*(hCover/prevTotalHeight)
+					h = minHeight*(h/prevTotalHeight)
+				}
+				
+				return MediaLayoutResult(width: Int(maxWidth),
+										 height: Int((hCover+h+gap).rounded()),
+										 columnSizes: [Int(w0.rounded()), Int(w1.rounded()), Int(maxWidth-w0.rounded()-w1.rounded())],
+										 rowSizes: [Int(hCover.rounded()), Int(h.rounded())],
+										 tiles: [
+											MediaLayoutResult.Tile(colSpan: 3, rowSpan: 1, startCol: 0, startRow: 0),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 0, startRow: 1),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 1, startRow: 1),
+											MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 2, startRow: 1)
+										 ])
+			} else {
+				// One on the left, three smaller ones on the right
```

Comment:
Does this mean 4 attachments will never be rendered as a 2x2 grid?

Reply:
Yes, it's always 1 bigger one and 3 smaller ones